plugins/backstage-plugin-coder/src/api/UrlSync.ts

```diff
@@ -0,0 +1,158 @@
+
+type UrlPrefixes = Readonly<{
+  proxyPrefix: string;
+  apiRoutePrefix: string;
+  assetsRoutePrefix: string;
+}>;
+
+export const defaultUrlPrefixes = {
+  proxyPrefix: `/api/proxy`,
+  apiRoutePrefix: '/api/v2',
+  assetsRoutePrefix: '', // Deliberately left as empty string
+} as const satisfies UrlPrefixes;
+
+export type UrlSyncSnapshot = Readonly<{
+  baseUrl: string;
+  apiRoute: string;
+  assetsRoute: string;
+}>;
+
+type Subscriber = SubscriptionCallback<UrlSyncSnapshot>;
+
+type ConstructorInputs = Readonly<{
+  urlPrefixes?: Partial<UrlPrefixes>;
+  apis: Readonly<{
+    discoveryApi: DiscoveryApi;
+    configApi: ConfigApi;
+  }>;
+}>;
+
+const proxyRouteReplacer = /\/api\/proxy.*?$/;
+
+type UrlSyncApi = Subscribable<UrlSyncSnapshot> &
+  Readonly<{
+    getApiEndpoint: () => Promise<string>;
+    getAssetsEndpoint: () => Promise<string>;
+    getCachedUrls: () => UrlSyncSnapshot;
+  }>;
+
+export class UrlSync implements UrlSyncApi {
+  // ConfigApi is literally only used because it offers a synchronous way to
+  // get an initial URL to use from inside the constructor. Should not be used
+  // beyond initial constructor call
+  private readonly configApi: ConfigApi;
+  private readonly discoveryApi: DiscoveryApi;
+  private readonly urlCache: StateSnapshotManager<UrlSyncSnapshot>;
+  private urlPrefixes: UrlPrefixes;
+
+  constructor(setup: ConstructorInputs) {
+    const { apis, urlPrefixes = {} } = setup;
+    const { discoveryApi, configApi } = apis;
+
+    this.discoveryApi = discoveryApi;
+    this.configApi = configApi;
+    this.urlPrefixes = { ...defaultUrlPrefixes, ...urlPrefixes };
+
+    const proxyRoot = this.getProxyRootFromConfigApi();
+    this.urlCache = new StateSnapshotManager<UrlSyncSnapshot>({
+      initialSnapshot: this.prepareNewSnapshot(proxyRoot),
+    });
+  }
+
+  private getProxyRootFromConfigApi(): string {
+    const baseUrl = this.configApi.getString(BASE_URL_KEY_FOR_CONFIG_API);
+    return `${baseUrl}${this.urlPrefixes.proxyPrefix}`;
+  }
+
+  private prepareNewSnapshot(newProxyUrl: string): UrlSyncSnapshot {
+    const { assetsRoutePrefix, apiRoutePrefix } = this.urlPrefixes;
+
+    return {
+      baseUrl: newProxyUrl.replace(proxyRouteReplacer, ''),
+      assetsRoute: `${newProxyUrl}${CODER_PROXY_PREFIX}${assetsRoutePrefix}`,
+      apiRoute: `${newProxyUrl}${CODER_PROXY_PREFIX}${apiRoutePrefix}`,
+    };
+  }
+
+  /* ***************************************************************************
+   * All public functions should be defined as arrow functions to ensure they
+   * can be passed around React without risk of losing their `this` context
+   ****************************************************************************/
+
+  getApiEndpoint = async (): Promise<string> => {
+    const proxyRoot = await this.discoveryApi.getBaseUrl(
+      PROXY_URL_KEY_FOR_DISCOVERY_API,
+    );
+
+    const newSnapshot = this.prepareNewSnapshot(proxyRoot);
+    this.urlCache.updateSnapshot(newSnapshot);
```

Comment:
Yeah, a subscription would be nice, and I think that would ultimately be the best design – I'm just not completely sure what the best way to implement it with the current APIs would be. `DiscoveryApi` doesn't have a subscription method

As far as the rounding, though, that was basically an aesthetic decision. If I look at my Coder dashboard right now:
![Screenshot 2024-04-30 at 11 03 45 AM](https://github.com/coder/backstage-plugins/assets/28937484/039469d5-4078-4cbc-a6b6-b72fc30c0c0d)

I think the images look a little sloppy, because there's no "aesthetic consistency" between them. They're not using the same colors, and by default, don't have the same shape, so the only thing giving them visual rhythm is the fact that they're placed consistently in the same spot. I think removing the rounding is going to make the UI look less polished

I've actually been meaning to open a PR for Coder core to update the CSS to get the emoji rounding issues fixed, too

Reply:
It might make sense to leave the being consistent up to the folks uploading the icons, as frustrating as that is.  If you check out the templates page, some of the non-emoji icons are non-circle shapes which I think would look weird or wrong if they were forcefully rounded.

I think this is why we have the padding to emojis, to avoid cutting them off, but regular images can have the same problem.  How do we tell the difference between an emoji or emoji-like image and one that can be rounded?

Reply:
Some would look better rounded though, maybe it makes sense to add a "Round this icon?" option to the template settings, otherwise folks have to manually round their icons which seems annoying.

Reply:
Yeah, good point – I could go either way, honestly

I feel like user avatars (and other images that are basically for UI purposes) kind of have the expectation that they can be morphed or changed into any number of shapes, and that if the image the user supplies doesn't look good with the styling the vendor/app uses, it's the user's responsibility to fix that. I can't remember the last time I've seen a modern site not use some amount of rounding with their "UI-ish" images. But I don't think workspaces/templates quite have that expectation

With the emoji, we at least know that they will always come from the same source (I don't think we give users the ability to define custom emoji), so they're all working with the same uniform constraints and can be offset the same consistent way

Realistically, though, most icons are going to have their main content in the center, so they naturally lend themselves to being rounded. I think I'd rather choose making 95% of icons look better, with 5% looking worse, over 100% looking "okay-ish"

Reply:
And the plugin does have logic to handle there not being any images, too – those use the same rounding
![Screenshot 2024-05-01 at 9 21 23 AM](https://github.com/coder/backstage-plugins/assets/28937484/2ce68b97-f285-4a74-a3a4-0cf015b67f7e)

If we get rid of the rounding, I guess we could keep the images rounded. If we remove the rounding across the board, though, I think the UI starts to look a little funny
![Screenshot 2024-05-01 at 9 22 11 AM](https://github.com/coder/backstage-plugins/assets/28937484/90ff484c-9739-4df6-9ff0-2890a4e19347)